include/sycl/algorithm/inner_product.hpp

```diff
@@ -39,43 +39,85 @@
+    auto input_buff2 = sycl::helpers::make_const_buffer(first2, last2);
+
+    return inner_product_sequential_sycl<typename ExecutionPolicy::kernelName>(exec.get_queue(), input_buff1,
+                                                                               input_buff2, value, size, op1, op2);
+  }
+};
+
+template <>
+struct InnerProductImpl<false> {
+  template <class ExecutionPolicy, class InputIt1, class InputIt2, class T,
+            class BinaryOperation1, class BinaryOperation2>
+  static T inner_product_sequential(ExecutionPolicy& exec, InputIt1 first1,
+                                    InputIt1 last1, InputIt2 first2, T value,
+                                    BinaryOperation1 op1, BinaryOperation2 op2) {
+    while (first1 != last1) {
+      value = op1(value, op2(*first1, *first2));
+      ++first1;
+      ++first2;
+    }
+    return value;
+  }
+};
+
+/* sequential inner_product.
+* @brief Returns the inner product of two vectors across the range [first1,
+* last1) by applying Functions op1 and op2.
 * Implementation of the command group that submits a for_each_n kernel,
 * According to Parallelism TS version n4507. Section 4.3.2
 * The kernel is implemented as a lambda.
 * @param ExecutionPolicy exec : The execution policy to be used
-* @param InputIterator first : Start of the range via a forward iterator
-* @param Size n : Specifies the number of valid elements
-* @param Function  f : No restrictions
-*/
-template <class ExecutionPolicy, class InputIt1, class InputIt2, class T>
-T inner_product_sequential(ExecutionPolicy &exec, InputIt1 first1,
-                           InputIt1 last1, InputIt2 first2, T value) {
-  while (first1 != last1) {
-    value = value + (*first1 * *first2);
-    ++first1;
-    ++first2;
-  }
-  return value;
-}
-
-/* inner_product.
-* @brief Returns the inner product of two vectors across the range [first1,
-* last1) by applying Functions op1 and op2. Implementation of the command group
-* that submits an inner_product kernel.
+* @param InputIterator first1 : Start of the range via a forward iterator
+* @param InputIterator last1 : End of the range via a forward iterator
+* @param InputIterator first2 : Start of the second vector via a forward iterator
+* @param Function  op1 : No restrictions
+* @param Function  op2 : No restrictions
 */
 template <class ExecutionPolicy, class InputIt1, class InputIt2, class T,
           class BinaryOperation1, class BinaryOperation2>
 T inner_product_sequential(ExecutionPolicy &exec, InputIt1 first1,
                            InputIt1 last1, InputIt2 first2, T value,
                            BinaryOperation1 op1, BinaryOperation2 op2) {
-  while (first1 != last1) {
-    value = op1(value, op2(*first1, *first2));
-    ++first1;
-    ++first2;
-  }
-  return value;
+  static constexpr bool UseSycl = has_get_buffer_method<InputIt1>::value && has_get_buffer_method<InputIt2>::value;
```

Comment:
why do you need this has get buffer? is this to identify the type of the iterator? can it not be detected using the iterator tag?

Reply:
Hum I didn't notice there was a buffer_iterator_tag I could use. This should be a bit cleaner if I use this but the trick will be the same except that I check if a typedef exists rather than a method.
Or maybe I could just use is_base_of SyclIterator? For some reason I thought I shouldn't include sycl_iterator.hpp but it should be fine right? I'll commit something after lunch.

Reply:
SyclIterator should be fine to include.